### PR TITLE
[Codegen] Add vector transfer + slice foldings in GenericVectorization

### PR DESCRIPTION
**Pull request by Max191**
Vectorizing a `linalg.copy` op can result in a sequence of
```
%extract = tensor.extract_slice %source
%read = vector.transfer_read %extract
%write = vector.transfer_read %dest
%insert = tensor.insert_slice %write into %dest
```
This sequence is folded by the transfer_write folder into
```
%extract = tensor.extract_slice %source
%insert = tensor.insert_slice %extract into %dest
```
In order to conserve the vector transfers, this PR adds folding patterns for vector transfer ops acting on insert/extract slice ops. This will fold the insert_slice into the transfer_write and the extract_slice into the transfer_read, and the vector transfers will not be folded.

This is turned off for the vector distribution pipeline because it causes distribution to fail in some cases.

Also removes `Codegen/LLVMGPU/test/conv_pipeline_test_rocm.mlir`, since it completes a TODO to remove the test after eliminating some undesired extra buffers.

### CODE REVIEW

**Comment by Max191** [comment]
needs tests, but I wanted to run benchmarks

EDIT: I see this is breaking several tests. I'll need to look into this more

File: compiler/src/iree/compiler/Codegen/LLVMGPU/test/conv_pipeline_test_rocm.mlir [hunk 1]
```diff
@@ -1,61 +0,0 @@
...
-      }
-    }
-  }
-}
-
-// TODO: This test reflects a bug related to how the convolution is bufferized
-// for the LLVMGPUVectorize pipeline, meaning these local memory allocations are
-// not desired. This test should be dropped once the extra buffers have been
-// eliminated.
-
```
Comment (Max191) [inline compiler/src/iree/compiler/Codegen/LLVMGPU/test/conv_pipeline_test_rocm.mlir:58]: The allocations seem to have disappeared after adding these patterns, so I am removing this test.

**Comment by Max191** [comment]
This needs to wait on a bug fix in https://github.com/llvm/llvm-project/pull/95020.

File: compiler/src/iree/compiler/Codegen/LLVMGPU/test/conv_pipeline_test_rocm.mlir (hunk 1 above)
Comment (hanhanW) [reply compiler/src/iree/compiler/Codegen/LLVMGPU/test/conv_pipeline_test_rocm.mlir:58]: Please add it to the PR description, so people don't need to track back to review comments to figure out why it is removed.

**Review by hanhanW — APPROVED** [review]
LGTM. We have to be careful of folding tensor.extract_slice and tensor.insert_slice ops away because some transformations heavily rely on these ops. It looks good on CPU side because there are no further tiling after vectorization. The rest is vector lowering.

I'm not sure how it works on GPU side wrt scf.forall lowering. I think it is okay because you provide an option to turn it off.

(cc @banach-space because of the changes of `vectorize_with_masking_and_hoist.mlir`.)

**Comment by Max191** [comment]
cc @benmxwl-arm since I just had to update a `generic_vectorizaton.mlir` test that you just added. It seems like a harmless change, but pinging just in case

**Comment by Max191** [comment]
Sorry, wrong account. @MacDue

**Comment by MacDue** [comment]
Looks harmless (it's still infers the right vector size :))

**Comment by ScottTodd** [comment]
Heads up - I see test failures in these modified tests after merge:
```
The following tests FAILED:
        141 - iree/compiler/Codegen/Common/test/generic_vectorization.mlir.test (Failed)
        217 - iree/compiler/Codegen/LLVMCPU/test/pipeline_pad_tests.mlir.test (Failed)
        242 - iree/compiler/Codegen/LLVMCPU/test/vectorize_with_masking_and_hoist.mlir.test (Failed)
```

Postsubmit CI jobs are running behind a 12h queue right now, so I kicked off a presubmit run at https://github.com/iree-org/iree/actions/runs/10014109365/job/27683126524?pr=17971 on https://github.com/iree-org/iree/pull/17971 to see if the CI machines can repro.

Local logs: https://gist.github.com/ScottTodd/cfceb22a41ca80257918d1d468b05ddb

**Comment by hanhanW** [comment]
Thanks for the heads-up! I'm taking a look and will provide a fix.

**Comment by ScottTodd** [comment]
Linux CI seemed to pass... I can try my local Windows at an earlier commit 🤔 

Do the logs give any clues?

**Comment by hanhanW** [comment]
> Linux CI seemed to pass... I can try my local Windows at an earlier commit 🤔

That's weird. I think it should generate deterministic IRs... compiling IREE on linux now

> 
> Do the logs give any clues?

I don't see issues, I need to repro it on my local machine.

**Comment by ScottTodd** [comment]
On Windows/MSVC, those tests passed at https://github.com/iree-org/iree/commit/4a1333183c845b5e76785be38cbc8528153c9c8b

so something in this range caused them to fail: https://github.com/iree-org/iree/compare/4a1333183c845b5e76785be38cbc8528153c9c8b...5b112cbc21046aaa028a05da62da1e3f8d19bc20

Ben suggests
> Usually is an issue with relying on undefined collection behavior

**Comment by ScottTodd** [comment]
Weeeeeird. I can't repro now, after trying to bisect through that history. Friday afternoon doing Friday afternoon things...

Sorry for the noise :P

**Comment by MacDue** [comment]
I've yet to track down exactly what's wrong, but this change breaks one of our internal tests, resulting in an output that is all NaNs. 
It also seems to prevent the hosting of reads/writes (which for SME tiles is _very_ costly). I'm trying to find a small reproducer now...

**Comment by MacDue** [comment]
Update: 

1. The bad hosting of reads/writes is real, so we may want to consider disabling (for the CPU backend?) this or at least having it off by default. However, I'm not quite sure I understand the motivation for this change (and where it needs to be enabled). Why do we want to "conserve the vector transfers"? 

2. The incorrect results don't seem to come _directly_ from this change, but instead come from some ~stack alignment~ addressing mode/calculation issues exposed  by this (which was quite the pain to track down :sweat_smile:)  See https://github.com/llvm/llvm-project/pull/100080.

**Comment by hanhanW** [comment]
> Why do we want to "conserve the vector transfers"?

Say that the original input is:

```mlir
%extract = tensor.extract_slice %source
%read = vector.transfer_read %extract
%write = vector.transfer_read %dest
%insert = tensor.insert_slice %write into %dest

**Comment by hanhanW** [comment]
> Why do we want to "conserve the vector transfers"?

Say that the original input is:

```mlir
%extract = tensor.extract_slice %source
%read = vector.transfer_read %extract
%write = vector.transfer_read %dest
%insert = tensor.insert_slice %write into %dest
```

Without the change, it becomes a `(tensor.extract_slice, tensor.insert_slice)` pair. There are no compute (or load/store) ops at tensor level. Then we bufferize the dispatch which gets

```mlir
%src = memref.subview (...)
%dest = memref.subview (...)
memref.copy (...)
```

There are no vector codes, which leads to scalar load/store. While we don't rely on LLVM vectorizer, they are scalar loads/stores without additional vectorization at buffer level.

The trend is moving the vectorization to tensor's world, so we want to get:

```mlir
vector.transfer_read %src
vector.transfer_write %read, %dest
```

This preserves the vector code after bufferization:

```mlir
%src = memref.subview (...)
%read = vector.transfer_read %src ...
%dest = memref.subview (...)
vector.transfer_write %read, %dest ...
```

Does it make sense?

I don't remember the hoisting issues that we hit on ARM path. Could you share the expected IR before and after the hoisting?

**Comment by MacDue** [comment]
So before this change we'd get:
<details>
<pre>
%15 = scf.for %arg4 = %c0 to %c352 step %c1 iter_args(%arg5 = %14) -> (tensor<?x?xf32>) {
	%extracted_slice_7 = tensor.extract_slice %extracted_slice[%arg4, 0] [1, %10] [1, 1] : tensor<352x?xf32> to tensor<1x?xf32>
	%extracted_slice_8 = tensor.extract_slice %extracted_slice_2[%arg4, 0] [1, %11] [1, 1] : tensor<352x?xf32> to tensor<1x?xf32>
	%extracted_slice_9 = tensor.extract_slice %arg5[0, 0] [%10, %11] [1, 1] : tensor<?x?xf32> to tensor<?x?xf32>
	%17 = vector.create_mask %c1, %10 : vector<1x[8]xi1>
	%18 = vector.transfer_read %extracted_slice_7[%c0, %c0], %cst_1, %17 {in_bounds = [true, true]} : tensor<1x?xf32>, vector<1x[8]xf32>
	%19 = vector.create_mask %c1, %11 : vector<1x[8]xi1>
	%20 = vector.transfer_read %extracted_slice_8[%c0, %c0], %cst_1, %19 {in_bounds = [true, true]} : tensor<1x?xf32>, vector<1x[8]xf32>
	%21 = vector.create_mask %10, %11 : vector<[8]x[8]xi1>
	%22 = vector.transfer_read %extracted_slice_9[%c0, %c0], %cst_1, %21 {in_bounds = [true, true]} : tensor<?x?xf32>, vector<[8]x[8]xf32>
	%23 = vector.create_mask %10, %11, %c1 : vector<[8]x[8]x1xi1>
	%24 = vector.mask %23 { vector.contract {indexing_maps = [affine_map<(d0, d1, d2) -> (d2, d0)>, affine_map<(d0, d1, d2) -> (d2, d1)>, affine_map<(d0, d1, d2) -> (d0, d1)>], iterator_types = ["parallel", "parallel", "reduction"], kind = #vector.kind<add>} %18, %20, %22 : vector<1x[8]xf32>, vector<1x[8]xf32> into vector<[8]x[8]xf32> } : vector<[8]x[8]x1xi1> -> vector<[8]x[8]xf32>
	%25 = vector.transfer_write %24, %extracted_slice_9[%c0, %c0], %21 {in_bounds = [true, true]} : vector<[8]x[8]xf32>, tensor<?x?xf32>
	%inserted_slice_10 = tensor.insert_slice %25 into %arg5[0, 0] [%10, %11] [1, 1] : tensor<?x?xf32> into tensor<?x?xf32>
	scf.yield %inserted_slice_10 : tensor<?x?xf32>
}
</pre>
</details>

And `iree-codegen-optimize-tensor-insert-extract-slices` would transform that to:
<details>
<pre>
%21 = vector.transfer_read %extracted_slice_5[%c0, %c0], %cst_1, %19 {in_bounds = [true, true]} : tensor<?x?xf32>, vector<[8]x[8]xf32>
%22 = scf.for %arg4 = %c0 to %c352 step %c1 iter_args(%arg5 = %21) -> (vector<[8]x[8]xf32>) {
	%extracted_slice_10 = tensor.extract_slice %extracted_slice[%arg4, 0] [1, %14] [1, 1] : tensor<352x?xf32> to tensor<1x?xf32>
	%25 = vector.transfer_read %4[%arg4, %arg0], %cst_1 {in_bounds = [true, true]} : tensor<352x128xf32>, vector<1x[8]xf32>
	%26 = vector.transfer_read %extracted_slice_10[%c0, %c0], %cst_1, %18 {in_bounds = [true, true]} : tensor<1x?xf32>, vector<1x[8]xf32>
	%27 = vector.mask %20 { vector.contract {indexing_maps = [affine_map<(d0, d1, d2) -> (d2, d0)>, affine_map<(d0, d1, d2) -> (d2, d1)>, affine_map<(d0, d1, d2) -> (d0, d1)>], iterator_types = ["parallel", "parallel", "reduction"], kind = #vector.kind<add>} %25, %26, %arg5 : vector<1x[8]xf32>, vector<1x[8]xf32> into vector<[8]x[8]xf32> } : vector<[8]x[8]x1xi1> -> vector<[8]x[8]xf32>
	scf.yield %27 : vector<[8]x[8]xf32>
}
%23 = vector.transfer_write %22, %extracted_slice_5[%c0, %c0], %19 {in_bounds = [true, true]} : vector<[8]x[8]xf32>, tensor<?x?xf32> 
</pre>
</details>

Which is super nice, and eventually means the initial `transfer_read` will become just a zero constant. 

It would hoist the pair:
```
%extracted_slice_9 = tensor.extract_slice %arg5[0, 0] [%10, %11] [1, 1] : tensor<?x?xf32> to tensor<?x?xf32>
%inserted_slice_10 = tensor.insert_slice %25 into %arg5[0, 0] [%10, %11] [1, 1] : tensor<?x?xf32> into tensor<?x?xf32>
```
Then the pair:
```
%22 = vector.transfer_read %extracted_slice_9[%c0, %c0], %cst_1, %21 {in_bounds = [true, true]} : tensor<?x?xf32>, vector<[8]x[8]xf32>
%25 = vector.transfer_write %24, %extracted_slice_9[%c0, %c0], %21 {in_bounds = [true, true]} : vector<[8]x[8]xf32>, tensor<?x?xf32>
```

Note: These are both easy to hoist as they're matching pairs to the same value (`%arg5` and `%extracted_slice_9` respectively).

---

Now we get:

<details>
<pre>
%16 = scf.for %arg4 = %c0 to %c352 step %c1 iter_args(%arg5 = %15) -> (tensor<?x?xf32>) {
	%extracted_slice_5 = tensor.extract_slice %arg5[0, 0] [%c8_vscale, %12] [1, 1] : tensor<?x?xf32> to tensor<?x?xf32>
	%18 = vector.transfer_read %4[%arg4, %arg0], %cst_1 {in_bounds = [true, true]} : tensor<352x128xf32>, vector<1x[8]xf32>
	%19 = vector.create_mask %c1, %12 : vector<1x[8]xi1>
	%20 = vector.transfer_read %5[%arg4, %arg2], %cst_1, %19 {in_bounds = [true, true]} : tensor<352x1xf32>, vector<1x[8]xf32>
	%21 = vector.create_mask %c8_vscale, %12 : vector<[8]x[8]xi1>
	%22 = vector.transfer_read %arg5[%c0, %c0], %cst_1, %21 {in_bounds = [true, true]} : tensor<?x?xf32>, vector<[8]x[8]xf32>
	%23 = vector.create_mask %c8_vscale, %12, %c1 : vector<[8]x[8]x1xi1>
	%24 = vector.mask %23 { vector.contract {indexing_maps = [affine_map<(d0, d1, d2) -> (d2, d0)>, affine_map<(d0, d1, d2) -> (d2, d1)>, affine_map<(d0, d1, d2) -> (d0, d1)>], iterator_types = ["parallel", "parallel", "reduction"], kind = #vector.kind<add>} %18, %20, %22 : vector<1x[8]xf32>, vector<1x[8]xf32> into vector<[8]x[8]xf32> } : vector<[8]x[8]x1xi1> -> vector<[8]x[8]xf32>
	%25 = vector.transfer_write %24, %extracted_slice_5[%c0, %c0], %21 {in_bounds = [true, true]} : vector<[8]x[8]xf32>, tensor<?x?xf32>
	%inserted_slice_6 = tensor.insert_slice %25 into %arg5[0, 0] [%c8_vscale, %12] [1, 1] : tensor<?x?xf32> into tensor<?x?xf32>
	scf.yield %inserted_slice_6 : tensor<?x?xf32>
}
</pre>
</details>

Here's no pairs `hoistSubsetAtIterArg()` knows how to hoist here. The sources no longer match up and there's uses in the loop blocking even hoisting the `tensor.extract_slice/insert_slice` pair).

So `iree-codegen-optimize-tensor-insert-extract-slices` fails to hoist anything (that matters at runtime):

<details>
<pre>
%21 = scf.for %arg4 = %c0 to %c352 step %c1 iter_args(%arg5 = %17) -> (tensor<?x?xf32>) {
	%extracted_slice_7 = tensor.extract_slice %arg5[0, 0] [%c8_vscale, %14] [1, 1] : tensor<?x?xf32> to tensor<?x?xf32>
	%23 = vector.transfer_read %4[%arg4, %arg0], %cst_1 {in_bounds = [true, true]} : tensor<352x128xf32>, vector<1x[8]xf32>
	%24 = vector.transfer_read %5[%arg4, %arg2], %cst_1, %18 {in_bounds = [true, true]} : tensor<352x1xf32>, vector<1x[8]xf32>
	%25 = vector.transfer_read %arg5[%c0, %c0], %cst_1, %19 {in_bounds = [true, true]} : tensor<?x?xf32>, vector<[8]x[8]xf32>
	%26 = vector.mask %20 { vector.contract {indexing_maps = [affine_map<(d0, d1, d2) -> (d2, d0)>, affine_map<(d0, d1, d2) -> (d2, d1)>, affine_map<(d0, d1, d2) -> (d0, d1)>], iterator_types = ["parallel", "parallel", "reduction"], kind = #vector.kind<add>} %23, %24, %25 : vector<1x[8]xf32>, vector<1x[8]xf32> into vector<[8]x[8]xf32> } : vector<[8]x[8]x1xi1> -> vector<[8]x[8]xf32>
	%27 = vector.transfer_write %26, %extracted_slice_7[%c0, %c0], %19 {in_bounds = [true, true]} : vector<[8]x[8]xf32>, tensor<?x?xf32>
	%inserted_slice_8 = tensor.insert_slice %27 into %arg5[0, 0] [%c8_vscale, %14] [1, 1] : tensor<?x?xf32> into tensor<?x?xf32>
	scf.yield %inserted_slice_8 : tensor<?x?xf32>
}
</pre>
</details>

This leaves very costly full-ZA tile reads/writes within in inner most loop of a matmul, which is really bad.

**Comment by banach-space** [comment]
Thanks for all the impressive digging, @MacDue ! Sadly this is hurting SME pretty bad :( (Sorry I didn't check earlier when @hanhanW pinged me).

@Max191, is [@tiled_linalg_copy](https://github.com/iree-org/iree/blob/1b60b62ed4b3f2a3c95f42b397889444c7df318c/compiler/src/iree/compiler/Codegen/Common/test/generic_vectorization.mlir#L406-L428) a good representation of what you wanted to achieve? I see 3 options here:
1. Find a way to preserve the behaviour of `@tiled_linalg_copy` while keeping nice hoisting that we used to have for SME  by moving `populateFoldTensorSubsetIntoVectorTransferPatterns` further down the compilation path,
2. Expose `earlySubsetTransferFolding` so that there's a user-facing flag that we could use in our compilation flow (I know that these are not popular),
3. Find a way to preserve the behaviour of `@tiled_linalg_copy` while keeping nice hoisting that we used to have for SME by means other than `populateFoldTensorSubsetIntoVectorTransferPatterns`. It's a bit counter-intuitive to me that these patterns are needed here TBH - what patterns eliminate `xfer_read`/`xfer_write` Ops that you want to preserve? (i.e., what's the root cause of the issue being addressed here?)

Thanks!

**Comment by qedawkins** [comment]
The root of the issue is that `linalg.copy` essentially vectorizes to a no-op, `transfer_write(transfer_read)` of the same indices. My opinion is that preventing such foldings shouldn't be a requirement for any lowering flow, i.e. if such a pair of transfers really do cancel, we should try to cancel them. Bufferization with produce a `memref.copy` later if the copy is material (i.e. between different memory spaces) and we can [vectorize such copies post bufferization](https://github.com/iree-org/iree/blob/main/compiler/src/iree/compiler/Codegen/Common/VectorizeMemrefCopy.cpp).

With that said, it looks like there is just a missing pattern that is blocking hoisting in the above example. I think we should do both of the following:

1. Revert this PR, it ideally shouldn't be needed (if someone gives a compelling reason why such a folding is a problem, we can think about relanding).
2. Fix or add the `insert_slice(transfer_write)` composition pattern because that's all that looks to be missing in the above example

```
%16 = scf.for %arg4 = %c0 to %c352 step %c1 iter_args(%arg5 = %15) -> (tensor) {
	%extracted_slice_5 = tensor.extract_slice %arg5[0, 0] [%c8_vscale, %12] [1, 1] : tensor to tensor
	...
	%22 = vector.transfer_read %arg5[%c0, %c0], %cst_1, %21 {in_bounds = [true, true]} : tensor, vector<[8]x[8]xf32>
        ...
	%25 = vector.transfer_write %24, %extracted_slice_5[%c0, %c0], %21 {in_bounds = [true, true]} : vector<[8]x[8]xf32>, tensor
	%inserted_slice_6 = tensor.insert_slice %25 into %arg5[0, 0] [%c8_vscale, %12] [1, 1] : tensor into tensor
	scf.yield %inserted_slice_6 : tensor
}
```

becomes the easily hoistable (and potentially cleaner IR)

```
%16 = scf.for %arg4 = %c0 to %c352 step %c1 iter_args(%arg5 = %15) -> (tensor) {
	...
	%22 = vector.transfer_read %arg5[%c0, %c0], %cst_1, %21 {in_bounds = [true, true]} : tensor, vector<[8]x[8]xf32>
        ...
	%25 = vector.transfer_write %24, %arg5[%c0, %c0], %21 {in_bounds = [true, true]} : vector<[8]x[8]xf32>, tensor
	scf.yield %25 : tensor
}
```

**Comment by MacDue** [comment]
Had a quick look a 2.:

The `insert_slice(transfer_write)` does not apply because the transfer_write is masked. So just looking at those two ops it may not be a legal replacement. I think you'd need to match `insert_slice(transfer_write(extract_slice))` and check both the insert + extract are from the same tensor.

**Comment by qedawkins** [comment]
I see, that makes sense. In that case I think the `insert_slice(transfer_write(extract_slice))` pattern would be worth writing then because the IR examples you showed above look reasonable to me. AFAICT this was a situation where we were just changing the order in which two different sets of folders were applied (`transfer_write(transfer_read)` vs `transfer(slice)`) and it would be good for all backends to handle both orders gracefully. That said if you want to send a revert, or disable it for CPU backends first, I would stamp.

> However, I'm not quite sure I understand the motivation for this change (and where it needs to be enabled). Why do we want to "conserve the vector transfers"?

To give a little more context, for GPU at some point (post bufferization) we really want to end up with these transfers, in particular because any time there is a `linalg.copy` on GPU it is being used to approximate data movement between memory spaces and/or layouts. Thus the transfers are there to vectorize the copy. With that said, `linalg.copy` is itself a no-op in the absence of the bufferization dialect, so this PR might have been a case of trying to preserve `linalg.copy` "nofold" semantics past the lifetime of the op, hence why I think reverting or toggling per backend makes sense.

**Comment by banach-space** [comment]
> That said if you want to send a revert, or disable it for CPU backends first, I would stamp.

https://github.com/iree-org/iree/pull/17997 :) I know that Ben is having a quick look at the pattern. I'm mostly keen to get our CI to a happier state.

From what you are saying, this PR was mostly to keep GPU and CPU paths consistent rather than fixing any specific CPU issue?

**Comment by qedawkins** [comment]
IIUC this PR was to change the behavior on GPU and opted to change the behavior on CPU also. The red flag for me is that this is disabled for `LLVMGPUVectorDistribute`, meaning we aren't even being consistent about it on the GPU side.

**Comment by MacDue** [comment]
Btw, I forgot to mention but when I took a look at the folds I spotted at least one upstream bug, which I reported here: [llvm/llvm-project#101708](https://github.com/llvm/llvm-project/issues/101708)

**Pull request merged**